PMM-6909 Rule templates API

artemgavrilov opened the pull request:
[PMM-6909](https://jira.percona.com/browse/PMM-6909)

- [ ] Build: https://github.com/Percona-Lab/pmm-submodules/pull/1271

- [ ] https://github.com/Percona-Lab/pmm-api-tests/pull/147
---
- [x] Tests passed.
- [x] Feature build pass.
- [ ] (Re)requested review.
- [x] Fix conflicts with target branch.
- [ ] Update API dependency.
- [ ] Update jira ticket description if needed.
- [ ] Attach screenshots/console output to confirm new behavior to jira ticket, if applicable.

artemgavrilov commented on models/template_model.go:
```diff
@@ -0,0 +1,132 @@
+}
+
+type FloatParam struct {
+	HasDefault bool    `json:"has_default"`
+	Default    float32 `json:"default"`
+
+	HasMin bool    `json:"has_min"`
+	Min    float32 `json:"min"`
+
+	HaxMax bool    `json:"hax_max"`
+	Max    float32 `json:"max"`
```
@AlekSi  param structure in API doesn't match param structure in `platform`. https://github.com/percona-platform/platform/blob/main/pkg/alert/parameter.go#L10
There is no way to specify only max in `alert.Rule`, moreover validation in `platform` requires full range and non-nil value.

AlekSi replied:
Let's ignore those fields and even remove from the model for now

AlekSi commented on models/template_model.go:
```diff
@@ -0,0 +1,117 @@
+	Params      Params   `reform:"params"`
+	For         Duration `reform:"for"`
+	Severity    string   `reform:"severity"`
+	Labels      Map      `reform:"labels"`
```
Please use the same approach as Agent:
* https://github.com/percona/pmm-managed/blob/85459abc452472c051f1070374d89dba2073c9be/models/agent_model.go#L74
* https://github.com/percona/pmm-managed/blob/85459abc452472c051f1070374d89dba2073c9be/models/agent_model.go#L140-L148

artemgavrilov replied:
Fixed

artemgavrilov commented on services/management/ia/templates_service.go:
```diff
@@ -171,19 +290,38 @@ func (svc *TemplatesService) ListTemplates(ctx context.Context, req *iav1beta1.L
+					Value: &iav1beta1.TemplateParam_Float{
+						Float: &iav1beta1.TemplateFloatParam{
+							HasDefault: true,           // TODO remove or fill with valid value.
+							Default:    float32(value), // TODO eliminate conversion.
```
In API values for `Float` have type `float32`, but in `saas.Rule` it's `float64`

artemgavrilov commented on services/management/ia/templates_service.go:
```diff
@@ -203,17 +341,80 @@ func (svc *TemplatesService) ListTemplates(ctx context.Context, req *iav1beta1.L
 }
 
 // UpdateTemplate updates existing template, previously created via API.
 func (svc *TemplatesService) UpdateTemplate(ctx context.Context, req *iav1beta1.UpdateTemplateRequest) (*iav1beta1.UpdateTemplateResponse, error) {
```
Should we after Create/Update refresh in-memory state, or it's up to the user (`reload` parameter in `ListTemplates` request)

artemgavrilov commented:
I'm still working on API tests, but that PR can be reviewed to save time.

artemgavrilov commented on services/management/ia/templates_service.go:
```diff
@@ -151,13 +250,33 @@ func (svc *TemplatesService) loadFile(ctx context.Context, file string) ([]saas.
+	}
+}
+
+func convertParamUnit(u string) iav1beta1.ParamUnit {
+	// TODO: check possible variants.
+	switch u {
+	case "%", "percentage":
+		return iav1beta1.ParamUnit_PERCENTAGE
+	default:
+		return iav1beta1.ParamUnit_PARAM_UNIT_INVALID
+	}
+}
```
There is no any validation for unit on `saas.Rule` side, there it should be just non-empty string. Here we have `enum`.

palash25 commented on models/template_model.go:
```diff
@@ -0,0 +1,144 @@
+	Summary     string        `reform:"summary"`
+	Tiers       Tiers         `reform:"tiers"`
+	Expr        string        `reform:"expr"`
+	Params      Params        `reform:"params"`
```
shouldn't this be a slice? https://github.com/percona/pmm/blob/PMM-2.0/api/managementpb/ia/templates.pb.go#L446

palash25 replied:
nevermind, I missed the type alias

palash25 commented on models/template_helpers.go:
```diff
@@ -0,0 +1,222 @@
+
+func checkUniqueTemplateName(q *reform.Querier, name string) error {
+	if name == "" {
+		panic("empty template name")
```
why do you choose to panic here instead of returing an error?

artemgavrilov replied:
Same as in https://github.com/percona/pmm-managed/blob/c2cd5d57da2f4dccfe7b2694e1f75e0f53125ebd/models/agent_helpers.go#L33

artemgavrilov replied:
I personally prefer not to panic) But it's common in `checkXXX` methods here.

AlekSi replied:
Panic there is used to indicate programmer's error – that function should not be called with empty ID; it is caller's responsibility to check that ID is not empty if it comes from the API call, etc. (in practice, that should be covered by protobuf validators, but please check that)

palash25 commented on models/template_helpers.go:
```diff
@@ -0,0 +1,222 @@
+}
+
+// FindTemplateByName finds template by name.
+func FindTemplateByName(q *reform.Querier, name string) (*Template, error) {
```
this function is same as the `checkUniqueTemplateName` except for the panic, why don't we drop the panic and reuse the code?

artemgavrilov replied:
Idea is that invocation `checkUniqueTemplateName` with empty rule name is the developer's mistake. For example, for agents, we create IDs on the server side, so they always should be present. In templates name provided by user, however, template `should` have a name, so we check it with `rule.Validate()` and notify user about bad request. So if `checkUniqueTemplateName` invoked with an empty name it's not something that can be handled, it's just serve bug.
On the other side, we have `FindTemplateByName` and here we have a different situation. Name for query provided by the user, so we should just notify him about the wrong request.

artemgavrilov replied:
Also `checkUniqueTemplateName` and `FindTemplateByName` have different errors/switch logic.

palash25 replied:
got it, thanks

palash25 commented on models/template_helpers.go:
```diff
@@ -0,0 +1,222 @@
+
+// FindTemplates returns saved notification rule templates.
+func FindTemplates(q *reform.Querier) ([]Template, error) {
+	structs, err := q.SelectAllFrom(TemplateTable, "")
```
nit: `structs` -> `rows`

artemgavrilov replied:
`structs` is consistent with other models.

https://github.com/percona/pmm-managed/blob/c2cd5d57da2f4dccfe7b2694e1f75e0f53125ebd/models/agent_helpers.go#L155

palash25 replied:
fine

palash25 commented on services/management/ia/templates_service.go:
```diff
@@ -338,18 +481,84 @@ func (svc *TemplatesService) ListTemplates(ctx context.Context, req *iav1beta1.L
+		DisallowInvalidRules:  true,
+	}
+
+	fmt.Println(req.Yaml)
```
i guess this was being used while debugging please remove

artemgavrilov replied:
Ah... removed

palash25 commented on services/management/ia/templates_service.go:
```diff
@@ -338,18 +481,84 @@ func (svc *TemplatesService) ListTemplates(ctx context.Context, req *iav1beta1.L
-func (svc *TemplatesService) CreateTemplate(ctx context.Context, req *iav1beta1.CreateTemplateRequest) (*iav1beta1.CreateTemplateResponse, error) {
-	return nil, status.Errorf(codes.Unimplemented, "method CreateTemplate not implemented")
+func (s *TemplatesService) CreateTemplate(ctx context.Context, req *iav1beta1.CreateTemplateRequest) (*iav1beta1.CreateTemplateResponse, error) {
+	pParams := &alert.ParseParams{
```
`pParams` -> `parseParams`

artemgavrilov replied:
Well, I don't know, the scope of that variable is a few lines. Does it require a more verbose name?

palash25 replied:
nah, its always better to have self-explantory names, but thats a nit

palash25 commented on services/management/ia/templates_service.go:
```diff
@@ -307,27 +427,50 @@ func (svc *TemplatesService) ListTemplates(ctx context.Context, req *iav1beta1.L
 			Labels:      r.Labels,
 			Annotations: r.Annotations,
-			Source:      iav1beta1.TemplateSource_TEMPLATE_SOURCE_INVALID, // TODO
+			Source:      r.Source,
```
nit: i guess this loop can be extracted out of the method, its getting too big, maybe move it to a `makeTemplate` function, something like that

artemgavrilov replied:
Well, I thought about that. In fact, this loop is almost everything that this method does, so extracting it will just produce a big function. Also, this code is not reusable now. IMHO better to leave it as is.

palash25 reviewed (approved):
just minor nits

artemgavrilov commented:
I updated the model to be more strict. Now it uses its own types for fields that can contain a limited amount of values. Also, I added converters for such types.

palash25 commented on services/management/ia/templates_service.go:
```diff
@@ -49,15 +53,23 @@ const (
 	l                    *logrus.Entry
 	builtinTemplatesPath string
 	userTemplatesPath    string
+	rulesFileDir         string
```
why this field? I don't see it being used in anyway for now. even in the tests you are just creating a temp dir and setting this field but what is the dir used for?

palash25 replied:
I just remembered  its supposed to be used in `dumpRule` here `	path := ruleFileDir + alertRule.Alert + ".yml"`

artemgavrilov replied:
Before that field tests dumped rule files in default dir. These files were caught by running managed instance it devcontainer and that caused the error. (It can't parse this format yet).
So I added that field to force tests to dump files in some temporary directory to eliminate side effects.

AlekSi commented on models/models.go:
```diff
@@ -127,8 +127,8 @@ func getCustomLabels(field []byte) (map[string]string, error) {
-// setCustomLabels encodes custom labels m to the Node/Service/Agent field.
-func setCustomLabels(m map[string]string, field *[]byte) error {
+// setStringMap encodes string map as model's byte array field.
+func setStringMap(m map[string]string, field *[]byte) error {
```
⚠️ `prepareLabels` validates that map keys are valid Prometheus labels. That's not a function any string map.

Let's call it getLabels (annotations are basically labels too)

artemgavrilov replied:
Done

AlekSi commented on models/database.go:
```diff
@@ -371,6 +371,28 @@ var databaseSchema = [][]string{
+			"for" BIGINT,
+			severity VARCHAR NOT NULL,
+			labels TEXT,
+			annotations TEXT,
```
Why TEXT and not JSONB? Can it be NOT NULL? Should it?

artemgavrilov replied:
Other models use `TEXT` for labels, so I used the same approach.
https://github.com/percona/pmm-managed/blob/c2cd5d57da2f4dccfe7b2694e1f75e0f53125ebd/models/database.go#L100

`alert.Rule` validation allows empty labels and annotations, so models follow common validness rules.

AlekSi replied:
If we are going to store JSON anyway, we probably may use JSONB and `{}` for empty set of labels. NULLs are less pleasant to work with

But no strong opinion from me there

artemgavrilov replied:
At first, it was JSONB, I change it to TEXT after that comment https://github.com/percona/pmm-managed/pull/541#discussion_r532482651

Will change to JSONB

artemgavrilov replied:
Ah, I remembered, we can use `JSOB` + Custom type with `Scan/Value` methods, **OR** we can use `TEXT` + `setStringMap`(will rename it) that returns byte slice.

artemgavrilov replied:
@AlekSi so we should choose something

AlekSi replied:
Ok, let's use `TEXT`

AlekSi commented on models/database.go:
```diff
@@ -371,6 +371,28 @@ var databaseSchema = [][]string{
+			labels TEXT,
+			annotations TEXT,
+			source VARCHAR NOT NULL,
+			yaml TEXT,
```
Why not NOT NULL?

artemgavrilov replied:
Was not sure that yaml source always will be present. Will change it.

AlekSi replied:
Well, it will be for all templates stored in the database. SaaS templates will not be in the database at all, for example.

But even in that case we could store an empty string. Again, just to avoid NULLs

artemgavrilov replied:
Done

artemgavrilov commented on services/management/ia/templates_service.go:
```diff
@@ -307,27 +462,50 @@ func (svc *TemplatesService) ListTemplates(ctx context.Context, req *iav1beta1.L
+					}
+
+					fp.HasMin = true      // TODO remove or fill with valid value.
+					fp.Min = float32(min) // TODO eliminate conversion.,
```
In API values for Float have type `float32`, but in saas.Rule it's `float64`

artemgavrilov commented on services/management/ia/templates_service.go:
```diff
@@ -338,18 +516,82 @@ func (svc *TemplatesService) ListTemplates(ctx context.Context, req *iav1beta1.L
 // CreateTemplate creates a new template.
-func (svc *TemplatesService) CreateTemplate(ctx context.Context, req *iav1beta1.CreateTemplateRequest) (*iav1beta1.CreateTemplateResponse, error) {
-	return nil, status.Errorf(codes.Unimplemented, "method CreateTemplate not implemented")
+func (s *TemplatesService) CreateTemplate(ctx context.Context, req *iav1beta1.CreateTemplateRequest) (*iav1beta1.CreateTemplateResponse, error) {
```
Should Create and Change template methods refresh in-memory state? Or it's up to the user to force it with `Reload` option in `ListTemplates` request

artemgavrilov commented on services/management/ia/templates_service.go:
```diff
@@ -151,15 +286,35 @@ func (svc *TemplatesService) loadFile(ctx context.Context, file string) ([]saas.
+	}
+}
+
+func convertParamUnit(u string) iav1beta1.ParamUnit {
+	// TODO: check possible variants.
+	switch u {
+	case "%", "percentage":
+		return iav1beta1.ParamUnit_PERCENTAGE
+	default:
+		return iav1beta1.ParamUnit_PARAM_UNIT_INVALID
+	}
 }
```
There is no any validation for unit on alert.Template side, it should be just non-empty string. Here we have enum.
